src/net.cpp

```diff
@@ -2636,6 +2633,9 @@ void CConnman::StopNodes()
     }
 
     // Close sockets
+    std::vector<CNode*> nodes;
+    std::list<CNode*> nodes_disconnected;
+    {
     LOCK(cs_vNodes);
```

Comment:
I think this lock can be held for much less time:

```c++
    // Delete peer connections
    std::vector<CNode*> nodes;
    {
        LOCK(cs_vNodes);
        nodes = std::move(vNodes);
        vNodes.clear();
    }

    for (CNode* pnode : nodes) {
        pnode->CloseSocketDisconnect();
        DeleteNode(pnode);
    }

    for (CNode* pnode : vNodesDisconnected) {
        DeleteNode(pnode);
    }
    vNodesDisconnected.clear();

    // Close listening sockets
    for (ListenSocket& hListenSocket : vhListenSocket) {
        if (hListenSocket.socket != INVALID_SOCKET) {
            if (!CloseSocket(hListenSocket.socket)) {
                LogPrintf("CloseSocket(hListenSocket) failed with error %s\n", NetworkErrorString(WSAGetLastError()));
            }
        }
    }
    vhListenSocket.clear();

    semOutbound.reset();
    semAddnode.reset();
```

Reply:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21563#issuecomment-818709445).